minidump/src/minidump.rs

```diff
@@ -116,6 +111,101 @@ pub trait MinidumpStream<'a>: Sized {
     fn read(bytes: &'a [u8], all: &'a [u8], endian: scroll::Endian) -> Result<Self, Error>;
 }
 
+/// Provides a unified interface for analyzing a [`MinidumpLinuxMaps`] and
+/// [`MinidumpMemoryInfoList`].
+///
+/// Currently an enum because there is no situation where you can have both,
+/// but this may change if the format evolves.
+#[derive(Debug, Clone)]
+pub enum UnifiedMemoryInfoList<'a> {
+    Maps(MinidumpLinuxMaps<'a>),
+    Info(MinidumpMemoryInfoList<'a>),
+}
+
+#[derive(Debug, Copy, Clone)]
+/// A [`UnifiedMemoryInfoList`] entry, providing metatadata on a region of
+/// memory in the crashed process.
+pub enum UnifiedMemoryInfo<'a> {
+    Map(&'a MinidumpLinuxMapInfo<'a>),
+    Info(&'a MinidumpMemoryInfo<'a>),
+}
+
+/// The contents of `/proc/self/mem` for the crashing process.
+///
+/// This is roughly equivalent in functionality to MinidumpMemoryInfoList.
+/// Use [`UnifiedMemoryInfoList`] to handle the two uniformly.
+#[derive(Debug, Clone)]
+pub struct MinidumpLinuxMaps<'a> {
+    /// The memory regions, in the order they were stored in the minidump.
+    regions: Vec<MinidumpLinuxMapInfo<'a>>,
+    /// Map from address range to index in regions. Use `MinidumpLinuxMaps::memory_info_at_address`.
+    regions_by_addr: RangeMap<u64, usize>,
+}
+
+/// A memory mapping entry for the process we are analyzing.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MinidumpLinuxMapInfo<'a> {
+    /// The first address this metadata applies to
+    pub base_address: u64,
+    /// The last address this metadata applies to
+    pub final_address: u64,
+
+    /// The kind of mapping
+    pub kind: MinidumpLinuxMapKind,
+
+    // FIXME: These could be bitflags but I'm not worried about it right now
+    pub is_read: bool,
+    pub is_write: bool,
+    pub is_exec: bool,
+    pub is_shared: bool,
+    pub is_private: bool,
+
+    // Fields in the format we ignore (not yet useful)
+    // * offset
+    // * dev
+    // * inode
+    _phantom: PhantomData<&'a u8>,
```

Comment:
I vaguely wanted to allow for a future where this struct is holding pointers into the minidump. Mostly because I was copying the MinidumpMemory code and so including lifetimes was the default. I could rip it all out if you want. The only thing that might be worth pointing into is the "filename" strings (currently I read them into actual Strings because I was feeling lazy).

Reply:
Oh, good point! I have a thought: how about you swap the `String`s for `Cow<'a, str>`, just make them always `Cow::Owned` with what you have now, and then that will let you use borrowed strings later, or if we get motivated to handle the invalid utf-8 case it will allow for that as well?